src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs

```diff
@@ -291,15 +291,54 @@ private bool StartCore(ProcessStartInfo startInfo)
                 }
             }
 
-            int childPid, stdinFd, stdoutFd, stderrFd;
+            int stdinFd = -1, stdoutFd = -1, stderrFd = -1;
             string[] envp = CreateEnvp(startInfo);
             string cwd = !string.IsNullOrWhiteSpace(startInfo.WorkingDirectory) ? startInfo.WorkingDirectory : null;
 
+            bool setCredentials = !string.IsNullOrEmpty(startInfo.UserName);
+            uint userId = 0;
+            uint groupId = 0;
+            if (setCredentials)
+            {
+                (userId, groupId) = GetUserAndGroupIds(startInfo);
+            }
+
             if (startInfo.UseShellExecute)
             {
+                if (!IsValidVerb(startInfo.Verb))
+                {
+                    throw new Win32Exception(Interop.Errors.ERROR_NO_ASSOCIATION);
+                }
+
+                // On Windows, UseShellExecute of executables and scripts causes those files to be executed.
+                // To achieve this on Unix, we check if the file is executable (x-bit).
+                // Some files may have the x-bit set even when they are not executable. This happens for example
+                // when a Windows filesystem is mounted on Linux. To handle that, treat it as a regular file
+                // when exec returns ENOEXEC (file format cannot be executed).
+                bool isExecuting = false;
+                filename = ResolveExecutableForShellExecute(startInfo.FileName);
```

Comment:
Consider changing ResolveExecutableForShellExecute to take a verb as an arg even if we only support only "open" right now. This way the verb check can be done over there and logic would be easier to modify in the future if we add other options.

Reply:
I'm going to leave this as is. I'll let the person that adds the next verb decide how to best change the code-flow, because it may depend on the verb.
@TSlivede I've restructured the verb check with the same reasoning in mind.

Reply:
The verb check looks fine to me now. It would require only small changes if someone wants to add other verbs, so I'm perfectly happy with the new implementation. (Even though the logic is IMHO still double negative, it's now much more clear and easier to change)